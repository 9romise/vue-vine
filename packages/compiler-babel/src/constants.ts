export const BARE_CALL_MACROS = [
  'vineExpose',
  'vineOptions',
  'vineStyle',
  'vineStyle.scoped',
] as const
export const VINE_MACROS = [
  'vineProp',
  'vineProp.optional',
  'vineProp.withDefault',
  'vineEmits',
  ...BARE_CALL_MACROS,
] as const
export const VINE_TAG_TEMPLATE_CALLER = [
  'vine',
  'css',
  'scss',
  'sass',
  'less',
  'stylus',
  'postcss',
] as const
export const SUPPORTED_CSS_LANGS = ['css', 'scss', 'sass', 'less', 'stylus', 'postcss'] as const
export const VUE_REACTIVITY_APIS = [
  'ref',
  'shallowRef',
  'computed',
  'reactive',
  'readonly',
  'watchEffect',
  'watchPostEffect',
  'watchSyncEffect',
  'watch',
  'triggerRef',
  'customRef',
  'shallowReactive',
  'shallowReadonly',
  'toRaw',
  'markRaw',
  'effectScope',
  'getCurrentScope',
  'onScopeDispose',
] as const
export const VUE_LIFECYCLE_HOOK_APIS = [
  'onMounted',
  'onUpdated',
  'onUnmounted',
  'onBeforeMount',
  'onBeforeUpdate',
  'onBeforeUnmount',
  'onErrorCaptured',
  'onRenderTracked',
  'onRenderTriggered',
  'onActivated',
  'onDeactivated',
  'onServerPrefetch',
] as const
export const TS_NODE_KINDS = [
  'as_expression',
  'type_assertion',
  'non_null_expression',
  'satisfies_expression',
] as const
export const BOOL_KINDS = ['true', 'false'] as const
