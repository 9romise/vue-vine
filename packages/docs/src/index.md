---
layout: home

hero:
  name: "Vue Vine"
  text: "Another style to write Vue."
  tagline: "Provide more flexibility for writing Vue components."
  image:
    src: /vine-logo.png
    alt: Vue Vine
  actions:
    - theme: brand
      text: Get Started
      link: /introduction/quick-start
    - theme: alt
      text: Why Vine
      link: /introduction/why

features:
  - title: Compatible <br> with SFC
    details: Vine is fully compatible with SFC, you can use it in your existing project.
  - title: Function component
    details: Write multiple components in one file, with all template features you love in Vue.js.
  - title: More uniform and intuitive type context
    details: Vine use TypeScript file everywhere so that type context would flow more naturally.
  - title: Simple and intuitive macros
    details: Extends design of Vue's macro mechanism, it's very concise and less mental burden.
---

<Recommend />

<Sponsors />
