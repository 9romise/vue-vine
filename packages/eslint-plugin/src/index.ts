import type { ESLint, Linter } from 'eslint'
import { version } from '../package.json'
import vineStyleIndent from './rules/vine-style-indent'

const plugin = {
  meta: {
    name: 'vue-vine',
    version,
  },
  rules: {
    'format-vine-style-indent': vineStyleIndent,
  },
} satisfies ESLint.Plugin

export default plugin

type RuleDefinitions = typeof plugin['rules']

export type RuleOptions = {
  [K in keyof RuleDefinitions]: RuleDefinitions[K]['defaultOptions']
}

export type Rules = {
  [K in keyof RuleOptions]: Linter.RuleEntry<RuleOptions[K]>
}
