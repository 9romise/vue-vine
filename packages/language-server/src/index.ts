import {
  createConnection,
  createServer,
  createSimpleProjectProviderFactory,
  createTypeScriptProjectProviderFactory,
  loadTsdkByPath,
} from '@volar/language-server/node'
import type * as ts from 'typescript'
import { create as createCssService } from 'volar-service-css'
import { create as createEmmetService } from 'volar-service-emmet'
import { create as createHtmlService } from 'volar-service-html'
import { create as createTypeScriptServices } from 'volar-service-typescript'

import type { VueCompilerOptions } from '@vue/language-core'
import { createParsedCommandLine, resolveVueCompilerOptions } from '@vue/language-core'
import { createVueVineLanguagePlugin } from '@vue-vine/language-service'

const debug = false

const connection = createConnection()
const server = createServer(connection)

connection.listen()

connection.onInitialize((params) => {
  const tsdk = loadTsdkByPath(
    params.initializationOptions.typescript.tsdk,
    params.locale,
  )

  return server.initialize(
    params,
    debug
      ? createTypeScriptProjectProviderFactory(
        tsdk.typescript,
        tsdk.diagnosticMessages,
      )
      : createSimpleProjectProviderFactory(),
    {
      getLanguagePlugins(env, projectContext) {
        let compilerOptions: ts.CompilerOptions = {}
        let vueCompilerOptions: VueCompilerOptions
        if (projectContext.typescript?.configFileName) {
          const { vueOptions, options } = createParsedCommandLine(tsdk.typescript, tsdk.typescript.sys, projectContext.typescript.configFileName)
          vueCompilerOptions = resolveVueCompilerOptions(vueOptions)
          compilerOptions = options
        }
        else {
          vueCompilerOptions = resolveVueCompilerOptions({})
        }
        return [createVueVineLanguagePlugin(tsdk.typescript, compilerOptions, vueCompilerOptions)]
      },
      getServicePlugins() {
        const plugins = [
          createHtmlService(),
          createCssService(),
          createEmmetService(),
        ]
        if (debug) {
          plugins.push(...createTypeScriptServices(tsdk.typescript))
        }
        return plugins;
      },
    },
  )
})

connection.onInitialized(server.initialized)
connection.onShutdown(server.shutdown)
