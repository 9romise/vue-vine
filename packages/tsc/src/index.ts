import { runTsc } from '@volar/typescript/lib/quickstart/runTsc'
import type { LanguagePlugin } from '@volar/language-core'
import { createVueVineLanguagePlugin } from '@vue-vine/language-service'
import { resolveVueCompilerOptions } from '@vue/language-core'

const removeEmitGlobalTypesRegexp = /^[^\n]*__VLS_globalTypesStart[\s\S]*__VLS_globalTypesEnd[^\n]*\n?$/gm

export function removeEmitGlobalTypes(dts: string) {
  return dts.replace(removeEmitGlobalTypesRegexp, '')
}

export function run() {
  const runExtensions = ['.vine.ts']
  const tscSdk = require.resolve('typescript/lib/tsc')
  const main = () => runTsc(
    tscSdk,
    runExtensions,
    (ts, runTscOptions) => {
      const languagePlugins: LanguagePlugin[] = []

      languagePlugins.push(
        createVueVineLanguagePlugin(
          ts,
          {
            compilerOptions: runTscOptions.options,
            vueCompilerOptions: resolveVueCompilerOptions({}),
            target: 'tsc',
          },
        ),
      )

      return {
        languagePlugins,
      }
    },
  )

  try {
    main()
  }
  catch (err) {
    console.error('[vue-vine-tsc]', err)
  }
}
