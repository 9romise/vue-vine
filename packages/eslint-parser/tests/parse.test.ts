import { describe, expect, it } from 'vitest'
import type { VineESLintParserOptions } from '../src/types'
import { runParse } from '../src/parse'

describe('vine ESLint parser test', () => {
  it('test variable reference', () => {
    const sampleSourceCode = `
function MyComp() {
  const v1 = 10
  const fn1 = () => {}

  return vine\`
    <div>{{ v1 }}</div>
    <button @click="fn1">Hi</button>
  \`
}`.trim()

    const parserOptions: VineESLintParserOptions = {
      ecmaVersion: 'latest',
      sourceType: 'module',
    }
    const { ast } = runParse(sampleSourceCode, parserOptions)
    expect(ast).toMatchSnapshot()
  })
})
