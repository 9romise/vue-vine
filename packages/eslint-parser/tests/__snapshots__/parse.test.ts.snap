// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vine ESLint parser test > test variable reference 1`] = `
{
  "body": [
    {
      "async": false,
      "body": {
        "body": [
          {
            "declarations": [
              {
                "definite": false,
                "id": {
                  "decorators": [],
                  "loc": {
                    "end": {
                      "column": 10,
                      "line": 2,
                    },
                    "start": {
                      "column": 8,
                      "line": 2,
                    },
                  },
                  "name": "v1",
                  "optional": false,
                  "parent": [Circular],
                  "range": [
                    28,
                    30,
                  ],
                  "type": "Identifier",
                  "typeAnnotation": undefined,
                },
                "init": {
                  "loc": {
                    "end": {
                      "column": 15,
                      "line": 2,
                    },
                    "start": {
                      "column": 13,
                      "line": 2,
                    },
                  },
                  "parent": [Circular],
                  "range": [
                    33,
                    35,
                  ],
                  "raw": "10",
                  "type": "Literal",
                  "value": 10,
                },
                "loc": {
                  "end": {
                    "column": 15,
                    "line": 2,
                  },
                  "start": {
                    "column": 8,
                    "line": 2,
                  },
                },
                "parent": [Circular],
                "range": [
                  28,
                  35,
                ],
                "type": "VariableDeclarator",
              },
            ],
            "declare": false,
            "kind": "const",
            "loc": {
              "end": {
                "column": 15,
                "line": 2,
              },
              "start": {
                "column": 2,
                "line": 2,
              },
            },
            "parent": [Circular],
            "range": [
              22,
              35,
            ],
            "type": "VariableDeclaration",
          },
          {
            "declarations": [
              {
                "definite": false,
                "id": {
                  "decorators": [],
                  "loc": {
                    "end": {
                      "column": 11,
                      "line": 3,
                    },
                    "start": {
                      "column": 8,
                      "line": 3,
                    },
                  },
                  "name": "fn1",
                  "optional": false,
                  "parent": [Circular],
                  "range": [
                    44,
                    47,
                  ],
                  "type": "Identifier",
                  "typeAnnotation": undefined,
                },
                "init": {
                  "async": false,
                  "body": {
                    "body": [],
                    "loc": {
                      "end": {
                        "column": 22,
                        "line": 3,
                      },
                      "start": {
                        "column": 20,
                        "line": 3,
                      },
                    },
                    "parent": [Circular],
                    "range": [
                      56,
                      58,
                    ],
                    "type": "BlockStatement",
                  },
                  "expression": false,
                  "generator": false,
                  "id": null,
                  "loc": {
                    "end": {
                      "column": 22,
                      "line": 3,
                    },
                    "start": {
                      "column": 14,
                      "line": 3,
                    },
                  },
                  "params": [],
                  "parent": [Circular],
                  "range": [
                    50,
                    58,
                  ],
                  "returnType": undefined,
                  "type": "ArrowFunctionExpression",
                  "typeParameters": undefined,
                },
                "loc": {
                  "end": {
                    "column": 22,
                    "line": 3,
                  },
                  "start": {
                    "column": 8,
                    "line": 3,
                  },
                },
                "parent": [Circular],
                "range": [
                  44,
                  58,
                ],
                "type": "VariableDeclarator",
              },
            ],
            "declare": false,
            "kind": "const",
            "loc": {
              "end": {
                "column": 22,
                "line": 3,
              },
              "start": {
                "column": 2,
                "line": 3,
              },
            },
            "parent": [Circular],
            "range": [
              38,
              58,
            ],
            "type": "VariableDeclaration",
          },
          {
            "argument": {
              "children": [
                {
                  "loc": {
                    "end": {
                      "column": 4,
                      "line": 10,
                    },
                    "start": {
                      "column": 14,
                      "line": 9,
                    },
                  },
                  "parent": [Circular],
                  "range": [
                    74,
                    79,
                  ],
                  "type": "VText",
                  "value": "
    ",
                },
                {
                  "children": [
                    {
                      "expression": Node {
                        "end": 15,
                        "loc": SourceLocation {
                          "end": {
                            "column": 14,
                            "line": 6,
                          },
                          "start": {
                            "column": 12,
                            "line": 6,
                          },
                        },
                        "name": "v1",
                        "parent": [Circular],
                        "range": [
                          87,
                          89,
                        ],
                        "start": 13,
                        "type": "Identifier",
                      },
                      "loc": {
                        "end": {
                          "column": 17,
                          "line": 10,
                        },
                        "start": {
                          "column": 9,
                          "line": 10,
                        },
                      },
                      "parent": [Circular],
                      "range": [
                        84,
                        92,
                      ],
                      "references": [
                        {
                          "id": Node {
                            "end": 15,
                            "loc": SourceLocation {
                              "end": {
                                "column": 14,
                                "line": 6,
                              },
                              "start": {
                                "column": 12,
                                "line": 6,
                              },
                            },
                            "name": "v1",
                            "parent": [Circular],
                            "range": [
                              87,
                              89,
                            ],
                            "start": 13,
                            "type": "Identifier",
                          },
                          "isTypeReference": undefined,
                          "isValueReference": undefined,
                          "mode": "r",
                        },
                      ],
                      "type": "VExpressionContainer",
                    },
                  ],
                  "endTag": {
                    "loc": {
                      "end": {
                        "column": 23,
                        "line": 14,
                      },
                      "start": {
                        "column": 17,
                        "line": 10,
                      },
                    },
                    "parent": [Circular],
                    "range": [
                      92,
                      98,
                    ],
                    "type": "VEndTag",
                  },
                  "loc": {
                    "end": {
                      "column": 23,
                      "line": 14,
                    },
                    "start": {
                      "column": 4,
                      "line": 14,
                    },
                  },
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "parent": [Circular],
                  "range": [
                    79,
                    98,
                  ],
                  "rawName": "div",
                  "startTag": {
                    "attributes": [],
                    "loc": {
                      "end": {
                        "column": 9,
                        "line": 10,
                      },
                      "start": {
                        "column": 4,
                        "line": 14,
                      },
                    },
                    "parent": [Circular],
                    "range": [
                      79,
                      84,
                    ],
                    "selfClosing": false,
                    "type": "VStartTag",
                  },
                  "type": "VElement",
                  "variables": [],
                },
                {
                  "loc": {
                    "end": {
                      "column": 4,
                      "line": 11,
                    },
                    "start": {
                      "column": 23,
                      "line": 10,
                    },
                  },
                  "parent": [Circular],
                  "range": [
                    98,
                    103,
                  ],
                  "type": "VText",
                  "value": "
    ",
                },
                {
                  "children": [
                    {
                      "loc": {
                        "end": {
                          "column": 27,
                          "line": 11,
                        },
                        "start": {
                          "column": 25,
                          "line": 11,
                        },
                      },
                      "parent": [Circular],
                      "range": [
                        124,
                        126,
                      ],
                      "type": "VText",
                      "value": "Hi",
                    },
                  ],
                  "endTag": {
                    "loc": {
                      "end": {
                        "column": 36,
                        "line": 15,
                      },
                      "start": {
                        "column": 27,
                        "line": 11,
                      },
                    },
                    "parent": [Circular],
                    "range": [
                      126,
                      135,
                    ],
                    "type": "VEndTag",
                  },
                  "loc": {
                    "end": {
                      "column": 36,
                      "line": 15,
                    },
                    "start": {
                      "column": 4,
                      "line": 15,
                    },
                  },
                  "name": "button",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "parent": [Circular],
                  "range": [
                    103,
                    135,
                  ],
                  "rawName": "button",
                  "startTag": {
                    "attributes": [
                      {
                        "directive": true,
                        "key": {
                          "argument": {
                            "loc": {
                              "end": {
                                "column": 18,
                                "line": 11,
                              },
                              "start": {
                                "column": 13,
                                "line": 11,
                              },
                            },
                            "name": "click",
                            "parent": [Circular],
                            "range": [
                              186,
                              191,
                            ],
                            "rawName": "click",
                            "type": "VIdentifier",
                          },
                          "loc": {
                            "end": {
                              "column": 18,
                              "line": 7,
                            },
                            "start": {
                              "column": 12,
                              "line": 11,
                            },
                          },
                          "modifiers": [],
                          "name": {
                            "loc": {
                              "end": {
                                "column": 13,
                                "line": 11,
                              },
                              "start": {
                                "column": 12,
                                "line": 11,
                              },
                            },
                            "name": "on",
                            "parent": [Circular],
                            "range": [
                              185,
                              186,
                            ],
                            "rawName": "@",
                            "type": "VIdentifier",
                          },
                          "parent": [Circular],
                          "range": [
                            111,
                            117,
                          ],
                          "type": "VDirectiveKey",
                        },
                        "loc": {
                          "end": {
                            "column": 24,
                            "line": 11,
                          },
                          "start": {
                            "column": 12,
                            "line": 11,
                          },
                        },
                        "parent": [Circular],
                        "range": [
                          111,
                          123,
                        ],
                        "type": "VAttribute",
                        "value": {
                          "expression": Node {
                            "end": 48,
                            "loc": SourceLocation {
                              "end": {
                                "column": 23,
                                "line": 7,
                              },
                              "start": {
                                "column": 20,
                                "line": 7,
                              },
                            },
                            "name": "fn1",
                            "parent": [Circular],
                            "range": [
                              119,
                              122,
                            ],
                            "start": 45,
                            "type": "Identifier",
                          },
                          "loc": {
                            "end": {
                              "column": 24,
                              "line": 11,
                            },
                            "start": {
                              "column": 19,
                              "line": 7,
                            },
                          },
                          "parent": [Circular],
                          "range": [
                            118,
                            123,
                          ],
                          "references": [
                            {
                              "id": Node {
                                "end": 48,
                                "loc": SourceLocation {
                                  "end": {
                                    "column": 23,
                                    "line": 7,
                                  },
                                  "start": {
                                    "column": 20,
                                    "line": 7,
                                  },
                                },
                                "name": "fn1",
                                "parent": [Circular],
                                "range": [
                                  119,
                                  122,
                                ],
                                "start": 45,
                                "type": "Identifier",
                              },
                              "isTypeReference": undefined,
                              "isValueReference": undefined,
                              "mode": "r",
                            },
                          ],
                          "type": "VExpressionContainer",
                        },
                      },
                    ],
                    "loc": {
                      "end": {
                        "column": 25,
                        "line": 11,
                      },
                      "start": {
                        "column": 4,
                        "line": 15,
                      },
                    },
                    "parent": [Circular],
                    "range": [
                      103,
                      124,
                    ],
                    "selfClosing": false,
                    "type": "VStartTag",
                  },
                  "type": "VElement",
                  "variables": [],
                },
                {
                  "loc": {
                    "end": {
                      "column": 2,
                      "line": 12,
                    },
                    "start": {
                      "column": 36,
                      "line": 11,
                    },
                  },
                  "parent": [Circular],
                  "range": [
                    135,
                    138,
                  ],
                  "type": "VText",
                  "value": "
  ",
                },
              ],
              "loc": {
                "end": {
                  "column": 2,
                  "line": 8,
                },
                "start": {
                  "column": 14,
                  "line": 5,
                },
              },
              "parent": [Circular],
              "range": [
                74,
                138,
              ],
              "type": "VTemplateRoot",
            },
            "loc": {
              "end": {
                "column": 3,
                "line": 8,
              },
              "start": {
                "column": 2,
                "line": 5,
              },
            },
            "parent": [Circular],
            "range": [
              62,
              139,
            ],
            "type": "ReturnStatement",
          },
        ],
        "loc": {
          "end": {
            "column": 1,
            "line": 9,
          },
          "start": {
            "column": 18,
            "line": 1,
          },
        },
        "parent": [Circular],
        "range": [
          18,
          141,
        ],
        "type": "BlockStatement",
      },
      "declare": false,
      "expression": false,
      "generator": false,
      "id": {
        "decorators": [],
        "loc": {
          "end": {
            "column": 15,
            "line": 1,
          },
          "start": {
            "column": 9,
            "line": 1,
          },
        },
        "name": "MyComp",
        "optional": false,
        "parent": [Circular],
        "range": [
          9,
          15,
        ],
        "type": "Identifier",
        "typeAnnotation": undefined,
      },
      "loc": {
        "end": {
          "column": 1,
          "line": 9,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
      "params": [],
      "parent": [Circular],
      "range": [
        0,
        141,
      ],
      "returnType": undefined,
      "type": "FunctionDeclaration",
      "typeParameters": undefined,
    },
  ],
  "comments": [],
  "loc": {
    "end": {
      "column": 1,
      "line": 9,
    },
    "start": {
      "column": 0,
      "line": 1,
    },
  },
  "parent": undefined,
  "range": [
    0,
    141,
  ],
  "sourceType": "module",
  "tokens": [
    {
      "loc": {
        "end": {
          "column": 8,
          "line": 1,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
      "range": [
        0,
        8,
      ],
      "type": "Keyword",
      "value": "function",
    },
    {
      "loc": {
        "end": {
          "column": 15,
          "line": 1,
        },
        "start": {
          "column": 9,
          "line": 1,
        },
      },
      "range": [
        9,
        15,
      ],
      "type": "Identifier",
      "value": "MyComp",
    },
    {
      "loc": {
        "end": {
          "column": 16,
          "line": 1,
        },
        "start": {
          "column": 15,
          "line": 1,
        },
      },
      "range": [
        15,
        16,
      ],
      "type": "Punctuator",
      "value": "(",
    },
    {
      "loc": {
        "end": {
          "column": 17,
          "line": 1,
        },
        "start": {
          "column": 16,
          "line": 1,
        },
      },
      "range": [
        16,
        17,
      ],
      "type": "Punctuator",
      "value": ")",
    },
    {
      "loc": {
        "end": {
          "column": 19,
          "line": 1,
        },
        "start": {
          "column": 18,
          "line": 1,
        },
      },
      "range": [
        18,
        19,
      ],
      "type": "Punctuator",
      "value": "{",
    },
    {
      "loc": {
        "end": {
          "column": 7,
          "line": 2,
        },
        "start": {
          "column": 2,
          "line": 2,
        },
      },
      "range": [
        22,
        27,
      ],
      "type": "Keyword",
      "value": "const",
    },
    {
      "loc": {
        "end": {
          "column": 10,
          "line": 2,
        },
        "start": {
          "column": 8,
          "line": 2,
        },
      },
      "range": [
        28,
        30,
      ],
      "type": "Identifier",
      "value": "v1",
    },
    {
      "loc": {
        "end": {
          "column": 12,
          "line": 2,
        },
        "start": {
          "column": 11,
          "line": 2,
        },
      },
      "range": [
        31,
        32,
      ],
      "type": "Punctuator",
      "value": "=",
    },
    {
      "loc": {
        "end": {
          "column": 15,
          "line": 2,
        },
        "start": {
          "column": 13,
          "line": 2,
        },
      },
      "range": [
        33,
        35,
      ],
      "type": "Numeric",
      "value": "10",
    },
    {
      "loc": {
        "end": {
          "column": 7,
          "line": 3,
        },
        "start": {
          "column": 2,
          "line": 3,
        },
      },
      "range": [
        38,
        43,
      ],
      "type": "Keyword",
      "value": "const",
    },
    {
      "loc": {
        "end": {
          "column": 11,
          "line": 3,
        },
        "start": {
          "column": 8,
          "line": 3,
        },
      },
      "range": [
        44,
        47,
      ],
      "type": "Identifier",
      "value": "fn1",
    },
    {
      "loc": {
        "end": {
          "column": 13,
          "line": 3,
        },
        "start": {
          "column": 12,
          "line": 3,
        },
      },
      "range": [
        48,
        49,
      ],
      "type": "Punctuator",
      "value": "=",
    },
    {
      "loc": {
        "end": {
          "column": 15,
          "line": 3,
        },
        "start": {
          "column": 14,
          "line": 3,
        },
      },
      "range": [
        50,
        51,
      ],
      "type": "Punctuator",
      "value": "(",
    },
    {
      "loc": {
        "end": {
          "column": 16,
          "line": 3,
        },
        "start": {
          "column": 15,
          "line": 3,
        },
      },
      "range": [
        51,
        52,
      ],
      "type": "Punctuator",
      "value": ")",
    },
    {
      "loc": {
        "end": {
          "column": 19,
          "line": 3,
        },
        "start": {
          "column": 17,
          "line": 3,
        },
      },
      "range": [
        53,
        55,
      ],
      "type": "Punctuator",
      "value": "=>",
    },
    {
      "loc": {
        "end": {
          "column": 21,
          "line": 3,
        },
        "start": {
          "column": 20,
          "line": 3,
        },
      },
      "range": [
        56,
        57,
      ],
      "type": "Punctuator",
      "value": "{",
    },
    {
      "loc": {
        "end": {
          "column": 22,
          "line": 3,
        },
        "start": {
          "column": 21,
          "line": 3,
        },
      },
      "range": [
        57,
        58,
      ],
      "type": "Punctuator",
      "value": "}",
    },
    {
      "loc": {
        "end": {
          "column": 8,
          "line": 5,
        },
        "start": {
          "column": 2,
          "line": 5,
        },
      },
      "range": [
        62,
        68,
      ],
      "type": "Keyword",
      "value": "return",
    },
    {
      "loc": {
        "end": {
          "column": 4,
          "line": 10,
        },
        "start": {
          "column": 14,
          "line": 9,
        },
      },
      "range": [
        74,
        79,
      ],
      "type": "HTMLWhitespace",
      "value": "
    ",
    },
    {
      "loc": {
        "end": {
          "column": 8,
          "line": 6,
        },
        "start": {
          "column": 4,
          "line": 14,
        },
      },
      "range": [
        79,
        83,
      ],
      "type": "HTMLTagOpen",
      "value": "div",
    },
    {
      "loc": {
        "end": {
          "column": 9,
          "line": 10,
        },
        "start": {
          "column": 8,
          "line": 6,
        },
      },
      "range": [
        83,
        84,
      ],
      "type": "HTMLTagClose",
      "value": "",
    },
    {
      "loc": {
        "end": {
          "column": 11,
          "line": 6,
        },
        "start": {
          "column": 9,
          "line": 10,
        },
      },
      "range": [
        84,
        86,
      ],
      "type": "VExpressionStart",
      "value": "{{",
    },
    Token {
      "end": 15,
      "loc": SourceLocation {
        "end": {
          "column": 14,
          "line": 6,
        },
        "start": {
          "column": 12,
          "line": 6,
        },
      },
      "range": [
        87,
        89,
      ],
      "start": 13,
      "type": "Identifier",
      "value": "v1",
    },
    {
      "loc": {
        "end": {
          "column": 17,
          "line": 10,
        },
        "start": {
          "column": 15,
          "line": 6,
        },
      },
      "range": [
        90,
        92,
      ],
      "type": "VExpressionEnd",
      "value": "}}",
    },
    {
      "loc": {
        "end": {
          "column": 22,
          "line": 6,
        },
        "start": {
          "column": 17,
          "line": 10,
        },
      },
      "range": [
        92,
        97,
      ],
      "type": "HTMLEndTagOpen",
      "value": "div",
    },
    {
      "loc": {
        "end": {
          "column": 23,
          "line": 14,
        },
        "start": {
          "column": 22,
          "line": 6,
        },
      },
      "range": [
        97,
        98,
      ],
      "type": "HTMLTagClose",
      "value": "",
    },
    {
      "loc": {
        "end": {
          "column": 4,
          "line": 11,
        },
        "start": {
          "column": 23,
          "line": 10,
        },
      },
      "range": [
        98,
        103,
      ],
      "type": "HTMLWhitespace",
      "value": "
    ",
    },
    {
      "loc": {
        "end": {
          "column": 11,
          "line": 7,
        },
        "start": {
          "column": 4,
          "line": 15,
        },
      },
      "range": [
        103,
        110,
      ],
      "type": "HTMLTagOpen",
      "value": "button",
    },
    {
      "loc": {
        "end": {
          "column": 19,
          "line": 7,
        },
        "start": {
          "column": 18,
          "line": 7,
        },
      },
      "range": [
        117,
        118,
      ],
      "type": "HTMLAssociation",
      "value": "",
    },
    {
      "loc": {
        "end": {
          "column": 20,
          "line": 7,
        },
        "start": {
          "column": 19,
          "line": 7,
        },
      },
      "range": [
        118,
        119,
      ],
      "type": "Punctuator",
      "value": "\\"",
    },
    Token {
      "end": 48,
      "loc": SourceLocation {
        "end": {
          "column": 23,
          "line": 7,
        },
        "start": {
          "column": 20,
          "line": 7,
        },
      },
      "range": [
        119,
        122,
      ],
      "start": 45,
      "type": "Identifier",
      "value": "fn1",
    },
    {
      "loc": {
        "end": {
          "column": 24,
          "line": 7,
        },
        "start": {
          "column": 23,
          "line": 7,
        },
      },
      "range": [
        122,
        123,
      ],
      "type": "Punctuator",
      "value": "\\"",
    },
    {
      "loc": {
        "end": {
          "column": 25,
          "line": 11,
        },
        "start": {
          "column": 24,
          "line": 7,
        },
      },
      "range": [
        123,
        124,
      ],
      "type": "HTMLTagClose",
      "value": "",
    },
    {
      "loc": {
        "end": {
          "column": 27,
          "line": 11,
        },
        "start": {
          "column": 25,
          "line": 11,
        },
      },
      "range": [
        124,
        126,
      ],
      "type": "HTMLText",
      "value": "Hi",
    },
    {
      "loc": {
        "end": {
          "column": 35,
          "line": 7,
        },
        "start": {
          "column": 27,
          "line": 11,
        },
      },
      "range": [
        126,
        134,
      ],
      "type": "HTMLEndTagOpen",
      "value": "button",
    },
    {
      "loc": {
        "end": {
          "column": 36,
          "line": 15,
        },
        "start": {
          "column": 35,
          "line": 7,
        },
      },
      "range": [
        134,
        135,
      ],
      "type": "HTMLTagClose",
      "value": "",
    },
    {
      "loc": {
        "end": {
          "column": 2,
          "line": 12,
        },
        "start": {
          "column": 36,
          "line": 11,
        },
      },
      "range": [
        135,
        138,
      ],
      "type": "HTMLWhitespace",
      "value": "
  ",
    },
    {
      "loc": {
        "end": {
          "column": 1,
          "line": 9,
        },
        "start": {
          "column": 0,
          "line": 9,
        },
      },
      "range": [
        140,
        141,
      ],
      "type": "Punctuator",
      "value": "}",
    },
    {
      "loc": {
        "end": {
          "column": 13,
          "line": 11,
        },
        "start": {
          "column": 12,
          "line": 11,
        },
      },
      "range": [
        185,
        186,
      ],
      "type": "Punctuator",
      "value": "@",
    },
    {
      "loc": {
        "end": {
          "column": 18,
          "line": 11,
        },
        "start": {
          "column": 13,
          "line": 11,
        },
      },
      "range": [
        186,
        191,
      ],
      "type": "HTMLIdentifier",
      "value": "click",
    },
  ],
  "type": "Program",
}
`;
